malariagen_data/anopheles.py

```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
-            self._setup_taxon_colors(plot_kwargs)
+        x = results["x"]
+        g123 = results["g123"]
 
-        debug("apply any user overrides")
-        plot_kwargs.update(kwargs)
+        return x, g123
 
-        debug("Create dendrogram with plotly")
-        # set labels as the index which we extract to reorder metadata
-        leaf_labels = np.arange(ht.shape[1])
-        # get the max distance, required to set xmin, xmax, which we need xmin to be slightly below 0
-        max_dist = _get_max_hamming_distance(
-            ht.T, metric="hamming", linkage_method=linkage_method
-        )
-        # noinspection PyTypeChecker
-        fig = create_dendrogram(
-            ht.T,
-            distfun=lambda x: _hamming_to_snps(x),
-            linkagefun=lambda x: linkage(x, method=linkage_method),
-            labels=leaf_labels,
-            color_threshold=0,
-            count_sort=count_sort,
-            distance_sort=distance_sort,
-        )
-        fig.update_traces(
-            hoverinfo="y",
-            line=dict(width=0.5, color="black"),
+    def _g123_gwss(
+        self,
+        contig,
+        sites,
+        site_mask,
+        window_size,
+        sample_sets,
+        sample_query,
+        min_cohort_size,
+        max_cohort_size,
+        random_seed,
+    ):
+        debug = self._log.debug
+        ds_snps = self.snp_calls(
+            region=contig,
+            sample_query=sample_query,
+            sample_sets=sample_sets,
+            site_mask=site_mask,
+            random_seed=random_seed,
         )
 
-        title_lines = []
-        if sample_sets is not None:
-            title_lines.append(f"sample sets: {sample_sets}")
-        if sample_query is not None:
-            title_lines.append(f"sample query: {sample_query}")
-        title_lines.append(f"genomic region: {region} ({ht.shape[0]} SNPs)")
-        title = "<br>".join(title_lines)
-
-        fig.update_layout(
-            width=width,
-            height=height,
-            title=title,
-            autosize=True,
-            hovermode="closest",
-            plot_bgcolor="white",
-            yaxis_title="Distance (no. SNPs)",
-            xaxis_title="Haplotypes",
-            showlegend=True,
-        )
+        if min_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples < min_cohort_size:
+                raise ValueError(
+                    f"not enough samples ({n_samples}) for minimum cohort size ({min_cohort_size})"
+                )
+        if max_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples > max_cohort_size:
+                rng = np.random.default_rng(seed=random_seed)
+                loc_downsample = rng.choice(
+                    n_samples, size=max_cohort_size, replace=False
+                )
+                loc_downsample.sort()
+                ds_snps = ds_snps.isel(samples=loc_downsample)
```

Comment:
Suggest to push this logic up into the snp_calls() function (which already has support for the cohort_size parameters).